APE17.rst

```diff
@@ -0,0 +1,482 @@
+.. _package-template: https://github.com/astropy/package-template
+
+Abstract
+--------
+
+The `astropy-helpers`_ package contains a collection of utilities intended to
+help build and distribute the core astropy package, as well as coordinated and
+affiliated packages in the Astropy ecosystem, and it is also used by packages
+outside of the ecosystem (at the time of writing, almost 400 repositories on
+GitHub appear to use astropy-helpers).
+
+The original APE proposing the creation of astropy-helpers (`APE 4`_) was
+accepted in June 2014. Since then, the state of the Python packaging ecosystem
+has changed significantly, with advances such as binary wheels that can be
+provided for all major platforms, tools such as `tox`_ and `setuptools_scm`_,
+and more recently support for `PEP 517`_ and `PEP 518`_ (in pip 19.0 or later),
+and many incremental updates and new features to `setuptools`_ (in setuptools
+30.3 or later). The purpose of this APE is to show how by relying on these
+tools, we can significantly simplify the packaging for astropy and all the
+affiliated packages and other packages using astropy-helpers, and remove the
+need for astropy-helpers altogether in pure-Python packages.
+
+Detailed description
+--------------------
+
+We now go through the various aspects of package infrastructure that
+astropy-helpers has solved in the past and how we can now address these instead
+using standard packaging tools.
+
+Changes common to all packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We first take a look at what changes are needed for all packages, regardless of
+whether they have compiled extensions or not. For pure-Python packages, if all
+these changes are made, astropy-helpers is no longer needed and can be removed
+from the repository.
+
+Running tests
+~~~~~~~~~~~~~
+
+Astropy-helpers provides the ability for anyone to run tests in the source
+distribution to do this with the::
+
+    python setup.py test
+
+command, which behind
+the scenes built the package, installed it into a temporary directory, installed
+pytest-astropy if needed, and ran the tests there.
+
+However, this can now be done more cleanly using `tox`_. Tox is a tool that
+makes is easy to install packages into a temporary clean Python environment
+along with the dependencies declared in setup.cfg, and run a set of custom
+commands - in a sense it does what the test command from astropy-helpers did,
+but in a more general way (including the use of an isolated test environment).
+For example, by putting the following into ``tox.ini``::
+
+    [testenv:test]
+    deps =
+      pytest-astropy
+    commands =
+      pytest aplpy
+
+developers will then easily be able to run the tests with::
+
+    tox -e test
+
+The benefit of using tox over just running pytest directly is that since behind
+the scenes it creates a source distribution of the package and installs it into
+a clean environment, it ensures that source distributions for the package are
```

Comment:
Does this mean it will need to rebuild the entire astropy every time you want to run tests in this way?

Reply:
Sorry, should have scrolled down two lines further.  So I think the answer is that developers need to manage the build themselves in order to do repetitive light-weight testing.

Reply:
I've now added some notes about this and what solutions exist to make this as easy as possible.